### ISSUE TITLE
`script::Builder::push_slice` is not beginner friendly. 

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#1746]
Pushing a script into a Script is a very common step that we need to do in wrapper segwit. sh(wsh), sh(wpkh). Here is how I had to do it in rust-miniscript.

```diff
+use bitcoin::script::{PushBytes, PushBytesBuf};
+use core::convert::TryFrom;
 use bitcoin::{script, Address, Network, ScriptBuf};
 
 
-                script::Builder::new()
-                    .push_script(&witness_script.to_v0_p2wsh())
-                    .into_script()
+                let push_bytes = PushBytesBuf::try_from(witness_script.into_bytes())
+                    .expect("Witness script is not too large");
+                script::Builder::new().push_slice(&push_bytes).into_script()
             }
             ShInner::Wpkh(ref wpkh) => {
                 let redeem_script = wpkh.script_pubkey();
-                script::Builder::new()
-                    .push_script(&redeem_script)
-                    .into_script()
+                let push_bytes: &PushBytes =
+                    <&PushBytes>::try_from(redeem_script.as_bytes()).unwrap();
+                script::Builder::new().push_slice(&push_bytes).into_script()

```
Doing a `wintess_script.as_bytes().try_into().unwrap()` does not work as rust still requires us to specify the exact that we want to try_into.
`multiple `impl`s satisfying `_: TryFrom<&[u8]>` found in the following crates: `bitcoin`, `core`: ...`

Importing a new type and reading it's documentation is a nice thing to do, but only to find out that we only fail when the total script size can exceed 2**32 which is exceedingly high. I propose we add the following methods that are more beginner friendly.

```rust

// We handle creating the pushbytes internally.
pub fn try_push_slice<S: AsRef<[u8]>>(sl: S) -> Result<Builder, Builder>;

pub fn push_slice_unchecked<S: AsRef<[u8]>>(sl: S) -> Builder;
```

Does not require the user to understand the inner workings of what pushbytes is and how it actually works. cc @Kixunil

**Comment by Kixunil** [comment]
In general, people should check the length upfront (also you shouldn't need to allocate `BuchBytesBuf` in your example) which is statically known in majority of cases. script-in-script is one of the rare exceptions and we should take a look into supporting it more directly.

That being said, I like `try_push_slice`. I'm not sure what's the idea behind `push_slice_unchecked` just to panic? People can use `unwrap` for it already. Breaking invariants instead doesn't sound great.

**Comment by sanket1729** [comment]
> also you shouldn't need to allocate BuchBytesBuf in your example

`PushBytesBuf` would take ownership of already allocated vector without doing any more allocations. I just highlighted a few ways to do it. 

> script-in-script is one of the rare exceptions and we should take a look into supporting it more directly.

We also need this for creating `ScriptSig` for legacy signatures or legacy multi-sig which also has the type `ScriptBuf`.

**Comment by sanket1729** [comment]
> That being said, I like try_push_slice. I'm not sure what's the idea behind push_slice_unchecked just to panic? 

Yeah, I don't like `push_slice_unchecked` anymore. I think we should support `try_push_slice`.

**Comment by Kixunil** [comment]
> legacy signatures or legacy multi-sig

Aren't those already statically known to be short? If yes it should be probably `struct ScriptSig(PushBytesBuf)` with conversion traits.

> unwrap which will almost certainly be the case for most applications.

I hope not, these may come from inputs which would create a DoS vulnerability.

**Comment by sanket1729** [comment]
>  If yes it should be probably struct ScriptSig(PushBytesBuf) with conversion traits.

We need some sort of builder to iteratively constructed the `ScriptBuf` for legacy multi-sig. Yes, but in theory all of them are known to be small. We could also have `push_ecdsa_sig(&EcdsaSig)`, `push_witness_script(&WScriptHash)` for some legacy functions.  

> I hope not, these may come from inputs which would create a DoS vulnerability.

True that. I certainly hope that applications that take script fragments input from untrusted parties are already something very weird :) 

Relatedly, we can also improve the 2^32 limit in PushBytes. Code form bitcoin core. It appears that all vectors can atmost have 2^25 elements. 

```C++
/**
 * The maximum size of a serialized object in bytes or number of elements
 * (for eg vectors) when the size is encoded as CompactSize.
 */
static constexpr uint64_t MAX_SIZE = 0x02000000;

/** Maximum amount of memory (in bytes) to allocate at once when deserializing vectors. */
static const unsigned int MAX_VECTOR_ALLOCATE = 5000000;
```

**Comment by Kixunil** [comment]
You already can push `EcdsaSig`: `script.push_slice(ecdsa_sig.serialize())`, same with `WScriptHash`: `script.push_slice(wscript_hash)`

`MAX_VECTOR_ALLOCATE` looks unrelated?

**Comment by sanket1729** [comment]
Note that ecdsa_sig returns a Vector. Not something of fixed size. 
The current implementations for PushBytes are for arrays of fixed size uptil 73. Which does not help if the we are trying to serialize a vector. For converting from vectors, we need to go via `PushBytes`.

> MAX_VECTOR_ALLOCATE looks unrelated?

Oops, extra copy paste :( . Editted away from previous comment.

**Comment by Kixunil** [comment]
> Note that ecdsa_sig returns a Vector

[It doesn't](https://docs.rs/bitcoin/latest/bitcoin/ecdsa/struct.Signature.html#method.serialize) I added `SerializedSignature` specifically to support this use case (also not allocating is nice), that's also why it impls `AsRef<PushBytes>` since it's guaranteed to be up to 73 bytes and commonly pushed into script.

**Comment by sanket1729** [comment]
> SerializedSignature specifically to support this us

Nice :) . Resolves 90% of the issue.

Still keeping this issue open for pushing ScriptBufs. If you are okay, I can add the `try_push_slice` method. 

> same with WScriptHash: script.push_slice(wscript_hash)

I intended  `push_witness_script(&WScriptHash)`  to compute the `ScriptBuf::new_v0_p2wsh` (`OP_PUSHBYTES32  <WSCRIPT_HASH> OP_EQUAL` ) and push that as slice. I can manually also push opcodes, but would like to avoid that.

**Comment by Kixunil** [comment]
Sure, feel free to add `try_push_slice` with appropriate documentation noting that many uses pushing common objects into script already work with `push_slice`.

Why would you ever push witness script rather than call `ScriptBuf::new_v0_p2wsh`? Witness script must not have anything before or after itself, so it looks more like a footgun.

**Comment by sanket1729** [comment]
Same use case. sh(wsh). The scriptsig needs to be `script.push_slice(&ScriptBuf::new_v0_p2wsh(...))`

**Comment by Kixunil** [comment]
Ah, I see. Feel free to add it too with doc where it's used and when to use the other one.

**Comment by apoelstra** [comment]
I think we should support `sh(wsh)` directly with a dedicated method. Then hopefully there are no more "sane" use cases for `try_push_slice`.

Then I think we should document `Builder::push_slice` to point at all the dedicated pushing methods that users might want instead of this method.
